### PR TITLE
perf: `std::sqrt` over `std::hypot` in Core

### PR DESCRIPTION
**Pull request by andiwand**
Replace `std::hypot` with `std::sqrt` in a couple of occasions in Core. This is numerically less accurate but much faster which is preferable for seeding and propagation.

blocked by
- https://github.com/acts-project/acts/pull/3701

---

This pull request focuses on optimizing mathematical calculations by replacing `std::hypot` with more efficient alternatives like `std::sqrt` and `Eigen`'s `norm` method. These changes aim to improve performance and maintain code consistency across various files.

Mathematical Optimization:

* [`Core/include/Acts/EventData/GenericFreeTrackParameters.hpp`](diffhunk://#diff-35a3d07a1eb05110c620e3cdbc38968456e66270c9f9460472f49e65d8b43acdL157-R162): Replaced `std::hypot` with `std::sqrt` for calculating `transverseMagnitude` and `magnitude`.
* [`Core/include/Acts/Propagator/EigenStepperDenseExtension.hpp`](diffhunk://#diff-3f30fd6031431b17dcd4e580cd73c202210417e80738cbcc76003b413be13c8dL111-R111): Replaced multiple instances of `std::hypot` with `std::sqrt` for various calculations, including `tKi`, `dtp1dl`, `dtp2dl`, `dtp3dl`, `dtp4dl`, and `energy`. [[1]](diffhunk://#diff-3f30fd6031431b17dcd4e580cd73c202210417e80738cbcc76003b413be13c8dL111-R111) [[2]](diffhunk://#diff-3f30fd6031431b17dcd4e580cd73c202210417e80738cbcc76003b413be13c8dL125-R125) [[3]](diffhunk://#diff-3f30fd6031431b17dcd4e580cd73c202210417e80738cbcc76003b413be13c8dL170-R179) [[4]](diffhunk://#diff-3f30fd6031431b17dcd4e580cd73c202210417e80738cbcc76003b413be13c8dL335-R359) [[5]](diffhunk://#diff-3f30fd6031431b17dcd4e580cd73c202210417e80738cbcc76003b413be13c8dL379-R385) [[6]](diffhunk://#diff-3f30fd6031431b17dcd4e580cd73c202210417e80738cbcc76003b413be13c8dL433-R439)
* [`Core/include/Acts/Propagator/StraightLineStepper.hpp`](diffhunk://#diff-15c3275a0d5f03483637d29160965a5e05496a788223d9e20f8b3ef225617c8aL338-R341): Replaced `std::hypot` with `std::sqrt` for calculating `dtds` and `prop_state.stepping.derivative`. [[1]](diffhunk://#diff-15c3275a0d5f03483637d29160965a5e05496a788223d9e20f8b3ef225617c8aL338-R341) [[2]](diffhunk://#diff-15c3275a0d5f03483637d29160965a5e05496a788223d9e20f8b3ef225617c8aL427-R428)
* [`Core/include/Acts/Seeding/EstimateTrackParamsFromSeed.hpp`](diffhunk://#diff-1d63df6364bcd57d6eae717e4c8dea355f07a36f5fca41b0e9fefdbf3224c0a0L245-R249): Replaced `std::hypot` with `std::sqrt` for calculating `invTanTheta` and `qOverPt`. [[1]](diffhunk://#diff-1d63df6364bcd57d6eae717e4c8dea355f07a36f5fca41b0e9fefdbf3224c0a0L245-R249) [[2]](diffhunk://#diff-1d63df6364bcd57d6eae717e4c8dea355f07a36f5fca41b0e9fefdbf3224c0a0L280-R279)
* [`Core/include/Acts/Utilities/VectorHelpers.hpp`](diffhunk://#diff-451e964d88f3d2d27ff239dade951924824ee73f885ef4044b5cea0861bfba40L76-R76): Replaced `std::hypot` with `Eigen`'s `norm` method for calculating the perpendicular distance.

Additional changes include similar optimizations in other files to ensure consistency and performance improvements across the codebase.

### CODE REVIEW

**Review by paulgessinger — COMMENTED** [review]
Sounds good. Did you measure this by any chance?

File: Core/include/Acts/Utilities/MathHelpers.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+#pragma once
+
+#include <cmath>
+
+namespace Acts {
+
+template <typename T>
+inline auto square(T x) {
+  return std::pow(x, 2);
```
Comment (CarloVarni) [inline Core/include/Acts/Utilities/MathHelpers.hpp:17]: why `std::pow` instead of `x*x` ?

Comment (andiwand) [reply Core/include/Acts/Utilities/MathHelpers.hpp:17]: No particular reason. I it should be equivalent

Comment (CarloVarni) [reply Core/include/Acts/Utilities/MathHelpers.hpp:17]: from some local tests it looks like `std::pow` is considerably more expensive then a simple multiplication

Comment (andiwand) [reply Core/include/Acts/Utilities/MathHelpers.hpp:17]: I think I discussed this once with @stephenswat and I believe a good optimizer should be able to substitute that. But you might be right in reducing the complexity by one step is more conservative. Will do

Comment (AJPfleger) [reply Core/include/Acts/Utilities/MathHelpers.hpp]: We had the discussion also here, and I also came. to the conclusion, that it shouldn't matter for squaring.

https://github.com/acts-project/acts/pull/3292#discussion_r1738601927

File: Core/include/Acts/Utilities/MathHelpers.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <cmath>
+
+namespace Acts {
+
+template <typename T>
+inline auto square(T x) {
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/MathHelpers.hpp:16]: I think `template` implies `inline`. 
Should we make these `constexpr`?

File: Core/include/Acts/Propagator/EigenStepperDenseExtension.hpp [hunk 3]
```diff
@@ -108,7 +109,7 @@ struct EigenStepperDenseExtension {
       // Evaluate k for the time propagation
       Lambdappi[0] = -qop[0] * qop[0] * qop[0] * g * energy[0] / (q * q);
       //~ tKi[0] = std::hypot(1, mass / initialMomentum);
-      tKi[0] = std::hypot(1, mass * qop[0]);
+      tKi[0] = hypot(1, mass * qop[0]);
```
[the remark this answers is not in the document]
Comment (andiwand) [reply Core/include/Acts/Propagator/EigenStepperDenseExtension.hpp:112]: I also had that concern. I think just adding `Acts::` might still creep in problems over time. Maybe we should just rename the function but I don't have a great idea to what

[the remark this answers is not in the document]
Comment (AJPfleger) [reply Core/include/Acts/Propagator/EigenStepperDenseExtension.hpp:112]: Yes, I can see, that it even gets removed sooner or later due to the redundancy.

I was thinking on maybe using the euclidean norm approach (before I saw your nice multi-argument implementation :3 )

Some naming ideas, that just popped up:
- `fastHypot`
- `multiHypot`
- `euclidNorm`
- `multiNorm`
- `fastNorm`

File: Tests/UnitTests/Core/Utilities/MathHelpersTests.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  const double x = std::pow(10, xExp);
+  const double y = std::pow(10, yExp);
+
+  const float fastFloat =
+      Acts::fastHypot(static_cast<float>(x), static_cast<float>(y));
+  const double fastDouble = Acts::fastHypot(x, y);
+
+  const double stdFloat =
+      std::hypot(static_cast<float>(x), static_cast<float>(y));
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Utilities/MathHelpersTests.cpp:37]:
```suggestion
  const float stdFloat =
      std::hypot(static_cast<float>(x), static_cast<float>(y));
```

File: Tests/UnitTests/Core/Utilities/QuickMathTests.cpp [hunk 5]
```diff
@@ -20,42 +20,42 @@ const auto expDist = bdata::random(
...
-BOOST_DATA_TEST_CASE(fastInverseSqrt, expDist ^ bdata::xrange(100), exp, i) {
+BOOST_DATA_TEST_CASE(quickInverseSqrt, expDist ^ bdata::xrange(100), exp, i) {
   (void)i;
 
   const double x = std::pow(10, exp);
 
-  const float fastFloat = Acts::fastInverseSqrt(static_cast<float>(x));
-  const double fastDouble = Acts::fastInverseSqrt(x);
+  const float quickFloat = Acts::quickInverseSqrt(static_cast<float>(x));
+  const double quickDouble = Acts::quickInverseSqrt(x);
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Utilities/QuickMathTests.cpp:29]: Just saw that we have here as well 2 lines below

`const double stdFloat = 1.0 / std::sqrt(static_cast<float>(x));`
which should be `const float`

Comment (andiwand) [reply Tests/UnitTests/Core/Utilities/QuickMathTests.cpp:29]: nice catch!

**Review by AJPfleger — APPROVED** [review]

File: Core/include/Acts/Utilities/QuickMath.hpp [hunk 6]
```diff
@@ -15,12 +15,12 @@
...
 namespace Acts {
 
-/// @brief Fast inverse square root function
+/// @brief Quick inverse square root function
 /// Taken from https://en.wikipedia.org/wiki/Fast_inverse_square_root
 /// @param x the number to calculate the inverse square root of
 /// @param iterations the number of newton iterations to perform
 template <typename T>
-constexpr T fastInverseSqrt(T x, int iterations = 1) noexcept {
+constexpr T quickInverseSqrt(T x, int iterations = 1) noexcept {
```
Comment (stephenswat) [inline Core/include/Acts/Utilities/QuickMath.hpp:23]: I know this is not covered in this PR, but my understanding is that, in modern compilers, this Quake (?) square root function isn't actually any faster than the `sqrt` builtin.

Comment (andiwand) [reply Core/include/Acts/Utilities/QuickMath.hpp:23]: https://github.com/acts-project/acts/pull/3701

**Review by AJPfleger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [4aa8311c088ba3ecd4a4cc75fe249a563847bc0d]
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/8285949)!

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472916)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472915) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_data18_13TeV_1000evt/pulls/3694/)|
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472914) | [:paperclip:](https://acts-athena-outputs.web.cern.ch/test_ttbarPU40_reco/pulls/3694/)|
|:red_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472913)||
|:red_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472912)||
|:red_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472911)||
|:red_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472910)||
|:red_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472909)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472908)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472907)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472906)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472905)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472904)||
|:red_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472903)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472902)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472901)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472900)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472899)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472898)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472897)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472896)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472895)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472894)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472893)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472892)||
|:red_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472891)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472890)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472889)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472888)||
|:red_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472887)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472886)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472885)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/44472884)||

**Comment by asalzburger** [comment]
I think that should go in.

**Comment by andiwand** [comment]
Just kicked off one more Athena CI run but the failure above seems unrelated

**Pull request merged**